### ISSUE TITLE
train issue:  Transform matrix for utterance <file> has bad dimension 13x13 versus feat dim 91

### ISSUE THREAD

**Issue by babua** [MontrealCorpusTools/Montreal-Forced-Aligner#281]
I'm trying to `mfa train` on my dataset, I have trained and aligned on this dataset successfully before. I have modified only the dictionary and trying to train again, and get this error at LDA stage. For what it's worth, I don't think dictionary is the problem, but I may have done a version upgrade in between, I honestly don't remember.

Some info:
```
version: 2.0.0a5
dirty: true
type: train_and_align
```

`mfa validate` runs successfully and finishes mono alignment. 

I find the error in `lda/log/acc.2.0.log`:

```
WARNING (transform-feats[5.5]:main():featbin/transform-feats.cc:110) Transform matrix for utterance <file> has bad dimension 13x13 versus feat dim 91
```

I get this warning for the whole corpus, then, at the end of `acc.2.0.log` :

```
LOG (transform-feats[5.5]:main():featbin/transform-feats.cc:161) Applied transform to 0 utterances; 10000 had errors.
LOG (gmm-acc-stats-ali[5.5]:main():gmmbin/gmm-acc-stats-ali.cc:112) Done 0 files, 0 with errors.
LOG (gmm-acc-stats-ali[5.5]:main():gmmbin/gmm-acc-stats-ali.cc:115) Overall avg like per frame (Gaussian only) = -nan over 0 frames.
LOG (gmm-acc-stats-ali[5.5]:main():gmmbin/gmm-acc-stats-ali.cc:123) Written accs.
WARNING (gmm-acc-stats-ali[5.5]:Close():util/kaldi-io.cc:515) Pipe apply-cmvn --utt2spk=ark:<home>/Documents/MFA/wav/corpus_data/subset_10000/utt2spk.0 scp:/home/onur/Documents/MFA/wav/corpus_data/subset_10000/cmvn.0.scp scp:<home>/Documents/MFA/wav/corpus_data/subset_10000/feats.0.scp ark:- | splice-feats --left-context=3 --right-context=3 ark:- ark:- | transform-feats <home>/Documents/MFA/wav/lda/lda.mat ark:- ark:- | had nonzero return status 256
ERROR (gmm-acc-stats-ali[5.5]:~SequentialTableReaderArchiveImpl():util/kaldi-table-inl.h:678) TableReader: error detected closing archive 'apply-cmvn --utt2spk=ark:<home>/Documents/MFA/wav/corpus_data/subset_10000/utt2spk.0 scp:<home>/Documents/MFA/wav/corpus_data/subset_10000/cmvn.0.scp scp:<home>/Documents/MFA/wav/corpus_data/subset_10000/feats.0.scp ark:- | splice-feats --left-context=3 --right-context=3 ark:- ark:- | transform-feats <home>/Documents/MFA/wav/lda/lda.mat ark:- ark:- |'
terminate called after throwing an instance of 'kaldi::KaldiFatalError'
  what():  kaldi::KaldiFatalError
```

Any help or guidance is appreciated. I think the MFCC's are being extracted 91-dimensional? But I have no idea why. 

Thank you in advance.

**Comment by mmcauliffe** [comment]
Can you try with the latest version of MFA (2.0.0a19) with the `--clean` flag?  I'm not too sure what's going wrong, but there have been a lot of updates to these systems, so it may have already been fixed.

**Comment by babua** [comment]
I tried with 2.0.0a19 and also with `--clean` flag, but getting the exact same error. But thanks for your reply!

**Comment by babua** [comment]
I did some more debugging and the issue seems to be about the splicing. I'll share here not asking for a solution but in case it helps debug something down the line. Unless there is something you'd like to add or comment, we can close the issue as I think the problem is out of scope for MFA devs.


I noticed that 91 = (3+1+3) * 13, which points at the concatenation of left and right context frames in LDA training along the wrong axis. I've verified this by setting different values in `config.yml` for `lda:features:splice_<left/right>_context`. Indeed it follows that pattern, for example if I set splice frames 2 each, the error message shows 65 instead of 91, 1 each, 39, and so on. 

I would guess this has something to do with some dependency version, but even when I make a completely fresh conda environment and install exactly like in the installation instructions, problem persists. Maybe it has something to do with my OpenBLAS installation or something at the system level.

**Comment by mmcauliffe** [comment]
Are there any errors in the previous logs?  It looks like this is throwing it at the second iteration, so I wonder if something happened the set up for creating the transform.  OpenBLAS errors would have likely shown up earlier, usually in the generation of MFCCs. 

Otherwise I'm a little stumped.  Any details about the corpus so I can try to reproduce?  Size of the corpus, sound file format, directory structure, sample rate, etc.

**Comment by babua** [comment]
Didn't check every single log but from a cursory look, I didn't see errors in the previous (mono, tri) logs. 

Also I managed to solve it in the end. The problem was indeed OpenBLAS. I was using MFA on this machine before the installation style changed, so I had OpenBLAS installed in a custom location (because I don't have sudo). I had some path variables set globally (LD_LIBRARY_PATH, BLAS, ATLAS). So even though I created a new conda environment with `openblas` , I think my path variables were overriding it. Simply `unset` ing those vars did the trick, ran the same exact `mfa` command line and worked without any problems. Sorry to take your time, but it was pretty sneaky and I was really at my wit's end. Thank you :)
